### ISSUE TITLE
Add a forward that gets called whenever an attribute gets added to a weapon

### ISSUE THREAD

**Issue by Zabaniya001** [nosoop/SM-TFCustAttr#24]
I'm well aware that it'd be best to avoid having a CW3-like approach where plugins have "boilerplate" to listen when an attribute gets added and from there they store everything in some global variables and work from there.

However, there are some cases where you'd want a passive ability of some sort or a gimmick and the only alternative is to detect when a weapon gets created and check if it has the attribute a few frames down the line.

**Comment by nosoop** [comment]
The trouble is that you have to consider the lifetime of the item as well; weapons can get transferred / dropped.  Covering every case is difficult, and the original intent of the API was to prevent certain assumptions.

I'm not against adding a sort of `forward void OnAttributeAdded(int entity)` that doesn't specify what was changed, I guess, though I expect that we'll be moving to proper attribute injection setups in the future.

**Comment by Zabaniya001** [comment]
> The trouble is that you have to consider the lifetime of the item as well; weapons can get transferred / dropped.

Yeah absolutely, that's expected. In one of my personal cases, I just need to detect if a weapon has a certain custom attribute so I can change some of its netprops. The other alternatives ( the one I mentioned in the original comment ) would just add unnecessary overhead.

> I'm not against adding a sort of forward void OnAttributeAdded(int entity) **that doesn't specify what was changed**

I can see why you would go with that approach ( to force people use the natives to check if said entity has the custom attribute ), but I think it'd just add unnecessary overhead since TF2CustAttr already knows what custom attribute is being added by a native before firing the hypothetical forward. I'm fine with it though, the natives *should* be fast enough to not care about it.

> though I expect that we'll be moving to proper attribute injection setups in the future.

I remember you mentioning that TF2CustAttr will ( probably ) eventually become just a front-end to the newer system of injecting attributes, but wasn't this project supposed to be gamedata-free?

**Comment by nosoop** [comment]
> I can see why you would go with that approach ( to force people use the natives to check if said entity has the custom attribute )

Also have to consider how to account for removals too.  `TF2CustAttr_UseKeyValues` would also need to iterate over keys.

> I remember you mentioning that TF2CustAttr will ( probably ) eventually become just a front-end to the newer system of injecting attributes, but wasn't this project supposed to be gamedata-free?

This project will probably remain gamedata-free; I might just leave it alone instead of integrating it into attribute injection.  Leave it to attributes to migrate themselves to use TF2Attributes and the managed injection system that I'm working on.

**Comment by Zabaniya001** [comment]
> Also have to consider how to account for removals too. TF2CustAttr_UseKeyValues would also need to iterate over keys.

Whilst it wouldn't be a bad addition, I don't think it's strictly necessary ( to have a forward that gets fired when an attribute gets removed ). Anywho, I haven't looked too much into it but generally attributes ( custom attributes that is, since you really can't use `TF2Attrib_RemoveByName` for them ) don't get removed except when a weapon gets destroyed, so you can probably fire it at that moment ( that's how I generally handle them at least ).

> This project will probably remain gamedata-free; I might just leave it alone instead of integrating it into attribute injection

You could probably just make a separate translation layer if really needed.

> managed injection system that I'm working on.

Nice, looking forward to it. Out of curiosity, are there going to be any major differences compared to [rafradek's](https://github.com/rafradek)
[TF2-Hidden-Attributes](https://github.com/rafradek/TF2-Hidden-Attributes)? I think you could just straight up integrate it to [tf2attributes](https://github.com/nosoop/tf2attributes) due to the ""similarity"" of the projects and use case, but it's just a thought.

**Comment by nosoop** [comment]
> I don't think it's strictly necessary ( to have a forward that gets fired when an attribute gets removed )

My imagined use case is, say, weapon / player models; someone that starts to rely on the forward-style of detecting real-time changes would have to deal with removals somehow (a "something changed" forward would work, a "this changed in particular" forward requires a bit more bookkeeping).  Even then it's prone to misuse / pitfalls, but I guess that's outside my scope of care.

If you can cover those cases, I guess you're welcome to submit a PR.

> Out of curiosity, are there going to be any major differences compared to [rafradek's](https://github.com/rafradek)
[TF2-Hidden-Attributes](https://github.com/rafradek/TF2-Hidden-Attributes)? I think you could just straight up integrate it to [tf2attributes](https://github.com/nosoop/tf2attributes) due to the ""similarity"" of the projects and use case, but it's just a thought.

The main difference is having a set of natives so plugins can register attributes themselves without having to worry about attribute ID collisions and fiddling with files.  The current implementation depends on C++ since it's way easier in native code to do insert / write operations with Valve's templated container types.
